direct/ingest-server/src/main/java/cz/o2/proxima/server/metrics/Metrics.java

```diff
@@ -116,38 +110,52 @@ static String getAttrNameForJMX(AttributeDescriptor<?> attr) {
 
   public static ApproxPercentileMetric sizeForAttribute(AttributeDescriptor<?> attr) {
     return getOrCreate(
-        String.format("%s_%s_size", attr.getEntity(), getAttrNameForJMX(attr)),
+        attr.getEntity() + "_" + getAttrNameForJMX(attr) + "_size",
         name ->
             ApproxPercentileMetric.of(
                 GROUP, name, Duration.ofHours(1).toMillis(), Duration.ofMinutes(5).toMillis()));
   }
 
-  public static void reportConsumerWatermark(String consumer, long watermark) {
+  public static void reportConsumerWatermark(
+      String consumer, long watermark, long elementTimestamp) {
     consumerWatermark(consumer, watermark);
     consumerWatermarkLag(consumer, watermark);
+    // Element timestamp is set to -1 for onIdle methods.
+    if (elementTimestamp >= 0) {
+      consumerWatermarkSkew(consumer, watermark, elementTimestamp);
+    }
   }
 
   private static void consumerWatermark(String consumer, long watermark) {
-
     GaugeMetric metric =
         getOrCreate(
-            String.format("%s_watermark", toJmxCompatibleConsumerName(consumer)),
+            toJmxCompatibleConsumerName(consumer) + "_watermark",
             name -> GaugeMetric.of(GROUP, name));
-
     consumerMetrics.put(consumer, metric);
     metric.increment(watermark);
   }
 
   private static void consumerWatermarkLag(String consumer, long watermark) {
     long lag = System.currentTimeMillis() - watermark;
     getOrCreate(
-            String.format("%s_watermark_lag", toJmxCompatibleConsumerName(consumer)),
+            toJmxCompatibleConsumerName(consumer) + "_watermark_lag",
             name ->
                 ApproxPercentileMetric.of(
                     GROUP, name, Duration.ofHours(1).toMillis(), Duration.ofMinutes(5).toMillis()))
         .increment(lag);
   }
 
+  private static void consumerWatermarkSkew(
+      String consumer, long watermark, long elementTimestamp) {
+    final long skew = Math.max(0, watermark - elementTimestamp);
+    getOrCreate(
+            toJmxCompatibleConsumerName(consumer) + "_watermark_skew",
```

Comment:
This probably is not "watermark skew". I'd suggest:
 - rename to `latecomer_lag`, or something similar
 - probably would be better to not calculate actual elements that are not latecomers at all, because otherwise it might be more than 99.9% zeros, which might make this metric useless

Reply:
I've changed the metric to calculate both positive and negative "skew". This gives us better understanding of what is actually happening.

Reply:
maybe `watermark_diff`, as this would not be focused solely on latecomers?

Reply:
:+1: